inc/saimacsec.h

```diff
@@ -0,0 +1,1285 @@
+     */
+    SAI_MACSEC_SC_ATTR_MACSEC_REPLAY_PROTECTION_WINDOW,
+
+    /**
+     * @brief MACsec SA list
+     * @type sai_object_list_t
+     * @flags READ_ONLY
+     * @objects SAI_OBJECT_TYPE_MACSEC_SA
+     */
+    SAI_MACSEC_SC_ATTR_SA_LIST,
+
+    /**
+     * @brief End of MACsec Secure Channel attributes
+     */
+    SAI_MACSEC_SC_ATTR_END,
+
+    /**
+     * @brief Custom range base value
+     */
+    SAI_MACSEC_SC_ATTR_CUSTOM_RANGE_START = 0x10000000,
+
+    /**
+     * @brief End of custom range base
+     */
+    SAI_MACSEC_SC_ATTR_CUSTOM_RANGE_END
+} sai_macsec_sc_attr_t;
+
+/**
+ * @brief MACsec Secure Channel counter IDs in sai_get_macsec_sc_stats() call
+ */
+typedef enum _sai_macsec_sc_stat_t
+{
+    /**
+     * @brief Control packets which are not secured (using MACsec uncontrolled port) -
+     * not in 802.1ae MIB
+     */
+    SAI_MACSEC_SC_STAT_CONTROL_PKTS,
+
+    /**
+     * @brief IEEE 802.1ae defined OutPktsUntagged or InPktsUntagged
+     */
+    SAI_MACSEC_SC_STAT_PKTS_UNTAGGED,
+
+    /**
+     * @brief Packets that have valid SCI, but the AN value does not have associated SA -
+     * not in 802.1ae MIB.
+     * Provides the sum of InPktsNotUsingSA for all invalid SAs for a SC
+     */
+    SAI_MACSEC_SC_STAT_SA_NOT_IN_USE,
+
+    /**
+     * @brief Control packets with SecTAG which are not secured (using MACsec
+     * uncontrolled port) - not in 802.1ae MIB
+     */
+    SAI_MACSEC_SC_STAT_TAGGED_CONTROL_PKTS,
+
+    /**
+     * @brief IEEE 802.1ae defined OutPktsTooLong.
+     * Valid for egress, always returns 0 for ingress.
+     */
+    SAI_MACSEC_SC_STAT_OUT_PKTS_TOO_LONG,
+
+    /**
+     * @brief IEEE 802.1ae defined InPktsNoTag.
+     * Valid for ingress, always returns 0 for egress.
+     */
+    SAI_MACSEC_SC_STAT_IN_PKTS_NO_TAG,
+
+    /**
+     * @brief IEEE 802.1ae defined InPktsBadTag.
+     * Valid for ingress, always returns 0 for egress.
+     */
+    SAI_MACSEC_SC_STAT_IN_PKTS_BAD_TAG,
+
+    /**
+     * @brief IEEE 802.1ae defined InPktsNoSci.
+     * Valid for ingress, always returns 0 for egress.
+     */
+    SAI_MACSEC_SC_STAT_IN_PKTS_NO_SCI,
+
+    /**
+     * @brief IEEE 802.1ae defined InPktsUnknownSci.
+     * Valid for ingress, always returns 0 for egress.
+     */
+    SAI_MACSEC_SC_STAT_IN_PKTS_UNKNOWN_SCI,
+
+    /**
+     * @brief IEEE 802.1ae defined InPktsOverrun.
+     * Valid for ingress, always returns 0 for egress.
+     */
+    SAI_MACSEC_SC_STAT_IN_PKTS_OVERRUN,
+} sai_macsec_sc_stat_t;
+
+/**
+ * @brief Attribute Id for sai_macsec_sa
+ */
+typedef enum _sai_macsec_sa_attr_t
+{
+    /**
+     * @brief Start of MACsec Secure Association attributes
+     */
+    SAI_MACSEC_SA_ATTR_START,
+
+    /**
+     * @brief MACsec direction
+     *
+     * @type sai_macsec_direction_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     */
+    SAI_MACSEC_SA_ATTR_MACSEC_DIRECTION = SAI_MACSEC_SA_ATTR_START,
+
```

Comment:
I had to remove SAI_MACSEC_SA_ATTR_SC_ID because, in conjunction with CREATE_AND_SET attribute SAI_MACSEC_SC_ATTR_ACTIVE_SA, it causes a loop of writable object_id (which causes a build error). But I forgot to make an associated change (make SAI_MACSEC_SC_ATTR_SA_LIST writable). But now I have a question (see below).

Context:
The normal sequence while creating a new egress SA is:
1. NOS creates a new SA and makes new SA active.  The driver invokes the hardware mechanism for atomically changing the SA.
2. NOS collect statistics for old SA.
3. NOS removes the old SA

Question: Which of the following options would be better for SAI:
Option-1 (fewer steps, explicit SAI_MACSEC_SA_ATTR_SC_ID, sequence-based state in driver):
Create-only attribute SAI_MACSEC_SA_ATTR_SC_ID and read-only attributes SAI_MACSEC_SC_ATTR_SA_LIST and SAI_MACSEC_SC_ATTR_ACTIVE_SA. The NOS only calls create_macsec_sa and the driver automatically updates SAI_MACSEC_SC_ATTR_SA_LIST and SAI_MACSEC_SC_ATTR_ACTIVE_SA.

Option-2 (more steps, no SAI_MACSEC_SA_ATTR_SC_ID, no sequence-based state in driver):
Writable attributes SAI_MACSEC_SC_ATTR_SA_LIST and SAI_MACSEC_SC_ATTR_ACTIVE_SA.  No SAI_MACSEC_SA_ATTR_SC_ID. The NOS has to perform 3 steps when creating SA:
create_macsec_sa
set_macsec_sc_attribute to update SAI_MACSEC_SC_ATTR_SA_LIST
set_macsec_sc_attribute to update SAI_MACSEC_SC_ATTR_ACTIVE_SA

Reply:
Meanwhile, I am assuming Option-1 for updating the header and documentation.